### PR TITLE
refactored cmd_print.c to use pj.c

### PR DESCRIPTION
**Pull request by deepakchethan**

### CODE REVIEW

File: libr/core/cmd_print.c [hunk 1]
```diff
@@ -2741,26 +2741,29 @@ static int cmd_print_blocks(RCore *core, const char *input) {
...
+
 			if ((as->block[p].flags)
 				|| (as->block[p].functions)
 				|| (as->block[p].comments)
 				|| (as->block[p].symbols)
 				|| (as->block[p].perm)
 				|| (as->block[p].strings)) {
-				r_cons_printf ("\"offset\":%"PFMT64u ",", at);
-				r_cons_printf ("\"size\":%"PFMT64u ",", piece);
+					pj_kn (pj, "offset", at);
```
Comment (ret2libc) [inline libr/core/cmd_print.c]: wrong indentation.

Comment (deepakchethan) [reply libr/core/cmd_print.c]: fixed it :)

**Comment by deepakchethan** [comment]
I have added a new method to pj.c library to print raw strings. The already present one was trying to escape utf8 and in doing so was adding additional `\`s.

File: libr/util/pj.c [hunk 2]
```diff
@@ -133,6 +133,24 @@ R_API PJ *pj_kd(PJ *j, const char *k, double d) {
...
 
+R_API PJ *pj_krs(PJ *j, const char *k, const char *v) {
+	if (j && v && *v) {
+		pj_k (j, k);
+		pj_rs (j, v);
+	}
+	return j;
+}
+
+R_API PJ *pj_rs(PJ *j, const char *k) {
```
Comment (radare) [inline libr/util/pj.c]: thats wrong

File: libr/util/pj.c [hunk 3]
```diff
@@ -133,6 +133,24 @@ R_API PJ *pj_kd(PJ *j, const char *k, double d) {
 	return j;
 }
 
+R_API PJ *pj_krs(PJ *j, const char *k, const char *v) {
```
Comment (radare) [inline libr/util/pj.c]: this too

File: libr/core/cmd_print.c [hunk 4]
```diff
@@ -5102,21 +5121,22 @@ static int cmd_print(void *data, const char *input) {
...
 				}
-
-				r_cons_printf ("{\"string\":");
+				pj_o (pj);
 				str = r_str_utf16_encode ((const char *) core->block, len);
-				r_cons_printf ("\"%s\"", str);
-				r_cons_printf (",\"offset\":%"PFMT64u, core->offset);
-				r_cons_printf (",\"section\":\"%s\"", vaddr == UT64_MAX? "unknown": section->name);
-				r_cons_printf (",\"length\":%d", len);
+				pj_krs (pj, "string", str);
```
Comment (radare) [inline libr/core/cmd_print.c]: just pass pj_ks and remove the r_str_utf16_encode which is wrong, the only thing you have to do is null terminate that core->block at the given len

Comment (deepakchethan) [reply libr/core/cmd_print.c]: On making the changes that you said. I am not getting the right output. The new output is as follows:

screen shot 2019-02-24 at 4 57 06 pm

**Comment by radare** [comment]
then we probably need to fix the json-utf8 call in the pj api, but it's a string, and there's only one type of string in json, adding a raw api to fill the string can be a root of problems

File: libr/util/pj.c (hunk 2 above)
Comment (deepakchethan) [reply libr/util/pj.c]: Removed it.

File: libr/util/pj.c (hunk 3 above)
Comment (deepakchethan) [reply libr/util/pj.c]: removed it

**Comment by deepakchethan** [comment]
> then we probably need to fix the json-utf8 call in the pj api, but it's a string, and there's only one type of string in json, adding a raw api to fill the string can be a root of problems
> […](#)

For now I reverted to the string implementation for that particular instance. I'll try to change the json-utf method in another pr. If that is okay with you.

File: libr/core/cmd_print.c [hunk 5]
```diff
@@ -5723,26 +5743,31 @@ static int cmd_print(void *data, const char *input) {
...
+
 						comma = ",";
 						// XXX: this only works in little endian
 						withref = 0;
 						if (core->print->hasrefs) {
 							char *rstr = core->print->hasrefs (core->print->user, val, true);
 							if (rstr && *rstr) {
 								char *ns = r_str_escape (rstr);
-								r_cons_printf (",\"ref\":\"%s\"}", *ns == ' ' ? ns + 1 : ns);
+								pj_ks (pj, "ref", *ns == ' ' ? ns + 1 : ns);
```
Comment (radare) [inline libr/core/cmd_print.c]: just use r_str_trim_ro (ns);

**Comment by radare** [comment]
i can merge after this fix, and yeah you can find a solution for the other problem later in a separate pr

File: libr/core/cmd_print.c (hunk 5 above)
Comment (deepakchethan) [reply libr/core/cmd_print.c]: Done

**Comment by ret2libc** [comment]
Thanks!

**Pull request merged**